csharp/src/Ice/Reference.cs

```diff
@@ -1825,5 +1555,296 @@ internal void Write(OutputStream ostr)
+                    response?.GetRetryPolicy(this) ?? exception!.GetRetryPolicy(request.IsIdempotent, sent);
+
+                // With the retry-policy OtherReplica we add the current connector to the list of excluded
+                // connectors and remove if from the list of connectors, this prevents the connector to be
+                // tried again during the current retry sequence.
+                if (retryPolicy == RetryPolicy.OtherReplica)
+                {
+                    if ((endpoints?[nextEndpoint] ?? connection?.Endpoint) is Endpoint endpoint)
+                    {
+                        excludedEndpoints ??= new();
+                        excludedEndpoints.Add(endpoint);
+                    }
+                    endpoints?.RemoveAt(nextEndpoint);
+                }
+
+                if (endpoints != null)
+                {
+                    if (connection == null && retryPolicy != RetryPolicy.OtherReplica)
+                    {
+                        // If connection establishment failed and the endpoint was not excluded, try the next
+                        // endpoint
+                        nextEndpoint = ++nextEndpoint % endpoints.Count;
+                    }
+
+                    if (endpoints.Count == 0 || nextEndpoint == 0)
+                    {
+                        // If the connector set is empty because all connectors has been excluded, or
+                        // nextConnector == 0, it indicates that we already tried all the connectors.
+                        if (cached)
+                        {
+                            // If the connectors were computed from cached endpoints, we clear the connectors to
+                            // trigger a new endpoint lookup.
+                            endpoints = null;
+                            nextEndpoint = 0;
+                        }
+                        else
+                        {
+                            // Otherwise we set triedAllConnectors to true to ensure further connection establishment
+                            // failures will now count as attempts (to prevent indefinitely looping if connection
+                            // establishment failure results in a retryable exception).
+                            triedAllEndpoints = true;
+                        }
+                    }
+                }
+
+                // Check if we can retry, we cannot retry if we have consumed all attempts, the current retry
+                // policy doesn't allow retries, the request was already released, there are no more endpoints
+                // or a fixed reference receives an exception with OtherReplica retry.
+
+                if (attempt == Communicator.InvocationMaxAttempts ||
+                    retryPolicy == RetryPolicy.NoRetry ||
+                    (sent && releaseRequestAfterSent) ||
+                    (triedAllEndpoints && endpoints != null && endpoints.Count == 0) ||
```

Comment:
Do we really need this condition if we no longer remove the endpoint from endpoints? If all the endpoints are excluded the retry will fail with NoEndpointException?

Reply:
I think we do if we remove them when we set `nextEndpoint` to 0